vllm/prefix.py

```diff
@@ -1,88 +1,79 @@
+        self.hash = hash(token_ids)
         assert self.length % block_size == 0
-        self.on_gpu = False
-        self.on_cpu = False
-        self.block_table: Optional[List[int]] = None
-        # a lock to prevent multiple sequence from calculating the same prefix
-        self.swap_to_gpu = False
+        self.block_table: Optional[BlockTable] = None
+
+    @property
+    def allocated(self) -> bool:
+        return self.block_table is not None
 
     def get_num_blocks(self) -> int:
         return self.length // self.block_size
 
     def get_block_numbers(self) -> List[int]:
         return [block.block_number for block in self.block_table]
 
-    def match(self, tokens: List[int]) -> bool:
-        return tokens[:self.length] == self.token_ids
-
-    # whether the prefix is on GPU or not
-    def get_status(self) -> bool:
-        return self.on_gpu
-
     def get_length(self) -> int:
         return self.length
 
+    def get_hash(self) -> int:
+        return self.hash
+
+    def set_block_table(self, block_table: BlockTable) -> None:
+        self.block_table = block_table.copy()
+
 
 class PrefixPool:
     """Manages all the prompt prefixes.
 
     Args:
         block_size: The block size of the executed model.
-    
+
     Attributes:
         prefixes: A list of all the prefixes.
-        prefixes_hash: Mapping from the hash of the prefix to the prefix id.
         block_size: The block size of the executed model.
     """
 
     def __init__(
         self,
         block_size: int,
     ) -> None:
-        self.prefixes = []
-        self.prefixes_hash = {}
+        self.prefixes: Dict[int, Prefix] = {}
         self.block_size = block_size
 
-    def add_prefix(self, token_ids: List[int]) -> Prefix:
-        prefix_hash = hash(tuple(token_ids))
-        assert prefix_hash not in self.prefixes_hash
-        # generate prefix_id
-        prefix_id = len(self.prefixes)
-        # create a new prefix
-        prefix = Prefix(prefix_id, token_ids, self.block_size)
-        self.prefixes.append(prefix)
-        self.prefixes_hash[prefix_hash] = prefix.prefix_id
-        return prefix
-
-    # use this first, if we already know from the application which part of the tokens are prefix.
-    def fixed_search(self, prefix_hash: int) -> Optional[Prefix]:
-        if prefix_hash not in self.prefixes_hash:
+    def _truncate_token_ids(self, token_ids: Sequence[int]) -> Tuple[int]:
+        new_length = len(token_ids) // self.block_size * self.block_size
+        return tuple(token_ids[:new_length])
+
+    def add_or_get_prefix(self, token_ids: Sequence[int]) -> Optional[Prefix]:
+        token_ids = self._truncate_token_ids(token_ids)
+        if len(token_ids) == 0:
+            # Prefix is empty.
             return None
-        prefix_id = self.prefixes_hash[prefix_hash]
-        return self.prefixes[prefix_id]
+        prefix = Prefix(token_ids, self.block_size)
+        prefix_hash = prefix.get_hash()
```

Comment:
You could implement the `__hash__` function in the `Prefix` class and by that eliminate some steps here.